samza-core/src/main/java/org/apache/samza/zk/ZkBarrierForVersionUpgrade.java

```diff
@@ -178,14 +202,15 @@ public void handleDataChange(String dataPath, Object data) {
       if (notAValidEvent())
         return;
 
-      zkUtils.unsubscribeDataChanges(barrierStatePath, this);
-      barrierListenerOptional.ifPresent(
-          zkBarrierListener -> zkBarrierListener.onBarrierStateChanged(barrierVersion, (State) data));
+      State barrierState = (State) data;
+      if (!Objects.equals(barrierState, State.NEW)) {
+        zkUtils.unsubscribeDataChanges(barrierStatePath, this);
+        barrierListenerOptional.ifPresent(zkBarrierListener -> zkBarrierListener.onBarrierStateChanged(barrierVersion, (State) data));
+      }
     }
 
     @Override
-    public void handleDataDeleted(String dataPath)
-        throws Exception {
+    public void handleDataDeleted(String dataPath) {
```

Comment:
Add a `getState()` API on the barrier that tests can assert when validating expected transitions. This should also clean-up tests a bit since it's un-necessary to expose the underlying ZkClient.

Reply:
I agree, we can eliminate redundancy in `getBarrierState()` calls in tests. I will add a method to test class itself,  we don't need to add it to `ZkBarrierForVersionUpgrade` class(other than tests it won't be used elsewhere).